promql/engine.go

```diff
@@ -452,13 +406,22 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *EvalStmt) (
 			return nil, err
 		}
 
+		// Turn matrix and vector types with protected metrics into
+		// model.* types.
+		switch v := val.(type) {
+		case vector:
+			val = v.value()
```

Comment:
Yes, all metrics that haven't been copied during evaluation are copied at the end and the `vector` and `matrix` types are turned into `model.Vector` and `model.Matrix`. Subsequent components (rules/ mostly) don't have to worry about modifying metrics.